File: eolymp/ford-bellman/ford-bellman.c
#include <stdio.h>
#include <stdlib.h>
#include <float.h>

#define INFINITY FLT_MAX

void solve() {
    int number_of_nodes, number_of_edges;
    scanf("%d %d", &number_of_nodes, &number_of_edges);

    int edges[number_of_edges][3];
    for (int i = 0; i < number_of_edges; i++) {
        scanf("%d %d %d", &edges[i][0], &edges[i][1], &edges[i][2]);
    }

    const int DEFAULT_DISCONNECTED_VALUE = 30000;
    float distances[number_of_nodes];
    for (int i = 0; i < number_of_nodes; i++) {
        distances[i] = INFINITY;
    }
    distances[0] = 0.0;

    const int RELAXATIONS = number_of_nodes - 1;
    for (int k = 0; k < RELAXATIONS; k++) {
        for (int i = 0; i < number_of_edges; i++) {
            int from = edges[i][0] - 1;
            int to = edges[i][1] - 1;
            int weight = edges[i][2];

            if (distances[from] != INFINITY && distances[from] + weight < distances[to]) {
                distances[to] = distances[from] + weight;
            }
        }
    }

    for (int i = 0; i < number_of_nodes; i++) {
        if (distances[i] != INFINITY) {
            printf("%d ", (int)distances[i]);
        } else {
            printf("%d ", DEFAULT_DISCONNECTED_VALUE);
        }
    }
    printf("\n");
}

int main() {
    solve();
    return 0;
}

File: eolymp/ford-bellman/ford-bellman.py
from math import inf


def solve():
    number_of_nodes, number_of_edges = list(map(int, input().split()))
    edges = []
    for _ in range(number_of_edges):
        edges.append(list(map(int, input().split())))

    DEFAULT_DISCONNECTED_VALUE = 30_000
    distances = [float("inf")] * number_of_nodes
    distances[0] = 0

    RELAXATIONS = number_of_nodes - 1
    for _ in range(RELAXATIONS):
        for from_, to_, weight in edges:
            if (
                distances[from_ - 1] != float("inf")
                and distances[from_ - 1] + weight < distances[to_ - 1]
            ):
                distances[to_ - 1] = distances[from_ - 1] + weight

    answer = [val if val != inf else DEFAULT_DISCONNECTED_VALUE for val in distances]
    print(*answer)


if __name__ == "__main__":
    solve()
